receiver/promtailreceiver/internal/logger.go

```diff
@@ -0,0 +1,151 @@
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/promtailreceiver/internal"
+
+import (
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"go.uber.org/zap"
+)
+
+const (
+	levelKey = "level"
+	msgKey   = "msg"
+	errKey   = "err"
+)
+
+// NewZapToGokitLogAdapter create an adapter for zap.Logger to log.Logger
+func NewZapToGokitLogAdapter(logger *zap.Logger) log.Logger {
+	// need to skip two levels in order to get the correct caller
+	// one for this method, the other for log
+	logger = logger.WithOptions(zap.AddCallerSkip(2))
+	return &zapToGokitLogAdapter{l: logger.Sugar()}
+}
+
+type zapToGokitLogAdapter struct {
+	l *zap.SugaredLogger
+}
+
+type logData struct {
+	level       level.Value
+	msg         string
+	otherFields []interface{}
+}
+
+func (w *zapToGokitLogAdapter) Log(keyvals ...interface{}) error {
+	// expecting key value pairs, the number of items need to be even
+	if len(keyvals)%2 == 0 {
+		// Extract log level and message and log them using corresponding zap function
+		ld := extractLogData(keyvals)
+		logFunc := levelToFunc(w.l, ld.level)
+		logFunc(ld.msg, ld.otherFields...)
+	} else {
+		// in case something goes wrong
+		w.l.Info(keyvals...)
+	}
+	return nil
+}
+
+func extractLogData(keyvals []interface{}) logData {
+	ld := logData{
+		level: level.InfoValue(), // default
+	}
+
+	for i := 0; i < len(keyvals); i += 2 {
+		key := keyvals[i]
+		val := keyvals[i+1]
+
+		if l, ok := matchLogLevel(key, val); ok {
+			ld.level = l
+			continue
+		}
+
+		if m, ok := matchLogMessage(key, val); ok {
+			ld.msg = m
+			continue
+		}
+
+		if err, ok := matchError(key, val); ok {
+			ld.otherFields = append(ld.otherFields, zap.Error(err))
+			continue
+		}
+
+		ld.otherFields = append(ld.otherFields, key, val)
+	}
+
+	return ld
+}
+
+// check if a given key-value pair represents go-kit log message and return it
+func matchLogMessage(key interface{}, val interface{}) (string, bool) {
+	if strKey, ok := key.(string); !ok || strKey != msgKey {
+		return "", false
+	}
+
+	msg, ok := val.(string)
+	if !ok {
+		return "", false
+	}
+	return msg, true
+}
+
+// check if a given key-value pair represents go-kit log level and return it
+func matchLogLevel(key interface{}, val interface{}) (level.Value, bool) {
+	strKey, ok := key.(string)
+	if !ok || strKey != levelKey {
+		return nil, false
+	}
+
+	levelVal, ok := val.(level.Value)
+	if !ok {
+		return nil, false
+	}
+	return levelVal, true
+}
+
+//revive:disable:error-return
```

Comment:
Do we need this?

Reply:
without this comment, golangci-lint throws an error: 
```
error-return: error should be the last type when returning multiple items (revive)
```

This code isn't mine actually, I copied it from prometheusreceiver/internal/logger.go and didn't refactor it.
I was thinking to move this code somewhere outside the components and import to promtail and prometheus receivers, but not sure if it is a good approach. 
If I move the logger somewhere (to internal directory maybe?) it will be there even if promtail and prometheus components are not present in the custom setup of collector contrib, so it will be an extra code user have to have that is not used anywhere